webapp/src/store/ProjectStore.ts

```diff
@@ -56,7 +56,7 @@ export class ProjectStore {
     }
 
     if (!this.data.languages[lang][id]) {
-      throw new MessageNotFound(lang, id);
+      this.data.languages[lang][id] = { text };
```

Comment:
Okay so here we are adding a text for an id, but the question is, where are we going to save this translation.

sourceFile hold the information about that, if we add a text without sourceFile file then where are we going to save this.

I know currently we are saving it in fixed file but this is wrong

Reply:
When @WULCAN gave me an introduction to Lyra we walked through the nature of this bug to give me an understanding of the system. 

We began discussing how your comment posed another level of complexity and how possibly solve it. 

For the the Zetkin file there is only one sourceFile to save to. The problem is when there is multiple files.
How can we create and save data at the correct location in the source code?

Based on the messageId we can search the file system for matching files and make a decision of which one to use. 

Example:

 zui.snackbar.info for `sv` does not have a value and gets a updated. 
 
We check the deepest path for `sv.yml`
 `locale/zui/snackbar/info/sv.yml`
 
There is no file here so we check a level shallower.
  `locale/zui/snackbar/sv.yml`

Neither a file at this location so we check another level shallower.
 `locale/zui/sv.yml`

Again no file matching, so we check the top possible level.
 `locale/sv.yml`
 
Here we find a `sv.yml` and can add a value for the messageId. If no file is present in either folder we could create one here.

What do you guys think?

Reply:
I think it's an interesting idea and it might work, but the one challenge I can think of is that when translating the first string from a particular file, that file may not yet exist. Imagine when adding a new language, like Finnish. When the language is first added, there are no `fi.yml` files, so you won't find one to put the new string in. So you would have to look for one of the other languages, e.g. Swedish, and put a new `fi.yml` next to `sv.yml`.

But then all of this is very specific for the multi-file YAML format, which is only one format that we want to be able to support. So putting this logical somewhere "central" might create too tight coupling.

I'm wondering if maybe we are thinking about this the wrong way. Maybe it should be the "message ID" that decides where translations go. That would work for old-school Zetkin translations (with multiple YAML files) because the message IDs are also YAML (the source of truth is the English YAML) and they live next to the translations so the path will be the same.

Reply:
Yes, if we want add a new language to https://github.com/zetkin/call.zetk.in/tree/master/locale with Lyra. The algorithm @a-jaxell and I suggested would fail to respect the layout and instead just add all new translations in a shared root translation file.

call.zetk.in is an application with message kind YAML where Lyra reads message definitions from `en.yml` or `en.yaml` instead of from `messageIds.ts`. New translation files should be siblings to that `en` file when this layout is used.

`app.zetkin.org` is an application with message kind TS where message definitions are read from `messageIds.ts` and those files are distributed in the source code hierarchy. One could conceivably attempt to recreate that hierarchy under `src/locale` for new languages or for messages without a translation, but that's not the layout used in `app.zetkin.org`. That repository is structured with shared translation files, in the root folder `src/locale`.

I think I can see a roadmap ahead resolving this with incrementally better features:

1. For projects with message kind TS, set `sourceFile` to `${lang}.yml` in the translation path root.
2. For projects with message kind YAML, set `sourceFile` to a sibling of the message-defining file.
3. Add support for more structures and file endings using new configuration in `lyra.yml`
4. Automatically detect these additional structures when not configured in `lyra.yml` 

Steps 3 and 4 can be guided by discovered need as we onboard more projects. There's no need to support every conceivable structure, we only need to support structures that are actually used with Lyra.

While 770deb302f0cc24bb256b815ebd99eb260272661 has no solution at all for this, it doesn't make the problem worse or harder to solve, and it fixes a very serious issue in Lyra: that you cannot save translations for new messages. Due to how `RepoGit.writeLangFiles` works on 770deb302f0cc24bb256b815ebd99eb260272661, the result will be equivalent to step 1 of the roadmap above. Can we merge this as is and extract the remaining plans to other issues, @amerharb ?

Reply:
yes, but it won't be like step 1.

in this PR it will be like "step 1" because sourceFile is kind of so far a meta data, but in my next PR (almost done but not tested) the saving will be based on this info, so my suggestion is: 

- make sourceFile in type MessageTranslation required again
- put value ${lang}.yaml in it here as you suggested, in this case it will be step 1 also but everyone is happy :)

Reply:
Step 2 of the roadmap only works if message-defining files are colocated with translation files. They are in `call.zetk.in` but not necessarily in all yaml message format projects.

Step 2 could be adapted to transpose the message path hierarchy into the translations hierarchy, or we could deny such layouts in Lyra until we discover need for them.